File: src/utils/read.zig
const std = @import("std");
const bytecode = @import("../common/bytecodes.zig");

pub fn read_u8(file: std.fs.File) ?u8 {
    var buffer: [1]u8 = undefined;
    const n = file.read(&buffer) catch {
        return null;
    };

    return if (n == 1) buffer[0] else null;
}

pub fn read_u16(file: std.fs.File) ?u16 {
    var buffer: [2]u8 = undefined;
    const n = file.read(&buffer) catch {
        return null;
    };

    if (n != 2) {
        return null;
    }

    return std.mem.readInt(u16, &buffer, .little);
}

pub fn read_u32(file: std.fs.File) ?u32 {
    var buffer: [4]u8 = undefined;
    const n = file.read(&buffer) catch {
        return null;
    };

    if (n != 4) {
        return null;
    }

    return std.mem.readInt(u32, &buffer, .little);
}

pub fn read_u64(file: std.fs.File) ?u64 {
    var buffer: [8]u8 = undefined;
    const n = file.read(&buffer) catch {
        return null;
    };

    if (n != 8) {
        return null;
    }

    return std.mem.readInt(u64, &buffer, .little);
}

pub fn read_f32(file: std.fs.File) ?f32 {
    return @bitCast(read_u32(file));
}

pub fn read_f64(file: std.fs.File) ?f64 {
    return @bitCast(read_u64(file));
}

pub fn read_string(file: std.fs.File, allocator: std.mem.Allocator) ?[]const u8 {
    var buffer: std.ArrayList(u8) = .empty;

    var c = read_u8(file);
    while (c != null and c != '\x00') {
        buffer.append(allocator, c.?) catch {
            // Skip
        };

        c = read_u8(file);
    }

    const result = buffer.toOwnedSlice(allocator) catch {
        return null;
    };

    return result;
}

pub fn read_bytecode(file: std.fs.File) ?bytecode.ApicaBytecode {
    if (read_u64(file)) |res| {
        return std.enums.fromInt(bytecode.ApicaBytecode, res);
    }
    return null;
}

pub fn read_entry_bytecode(file: std.fs.File) ?bytecode.ApicaEntrypointBytecode {
    if (read_u64(file)) |res| {
        return std.enums.fromInt(bytecode.ApicaEntrypointBytecode, res);
    }
    return null;
}

pub fn read_builtin_func_bytecode(file: std.fs.File) ?bytecode.ApicaBuiltinFuncCallBytecode {
    if (read_u64(file)) |res| {
        return std.enums.fromInt(bytecode.ApicaBuiltinFuncCallBytecode, res);
    }
    return null;
}

pub fn read_type_bytecode(file: std.fs.File) ?bytecode.ApicaTypeBytecode {
    if (read_u64(file)) |res| {
        return std.enums.fromInt(bytecode.ApicaTypeBytecode, res);
    }
    return null;
}

File: src/utils/read.cpp
#include "utils/read.hpp"
#include "utils/errors.hpp"
#include "systems/logger.hpp"

std::optional<uint8_t> utils::readU8(FILE *file) {
    int result = fgetc(file);
    if (result == EOF) {
        systems::LoggerSystem::getInstance().systemLognError(std::string(utils::RDR_ERROR_READ_U8));
        return std::nullopt;
    }
    
    return static_cast<uint8_t>(result);
}

std::optional<uint16_t> utils::readU16(FILE *file) {
    uint16_t result;
    if (fread(&result, sizeof(uint16_t), 1, file) != 1) {
        systems::LoggerSystem::getInstance().systemLognError(std::string(utils::RDR_ERROR_READ_U16));
        return std::nullopt;
    }
    
    return result;
}

std::optional<uint32_t> utils::readU32(FILE *file) {
    uint32_t result;
    if (fread(&result, sizeof(uint32_t), 1, file) != 1) {
        systems::LoggerSystem::getInstance().systemLognError(std::string(utils::RDR_ERROR_READ_U32));
        return std::nullopt;
    }
    
    return result;
}

std::optional<uint64_t> utils::readU64(FILE *file) {
    uint64_t result;
    if (fread(&result, sizeof(uint64_t), 1, file) != 1) {
        systems::LoggerSystem::getInstance().systemLognError(std::string(utils::RDR_ERROR_READ_U64));
        return std::nullopt;
    }
    
    return result;
}

std::optional<std::string> utils::readString(FILE *file) {
    std::string result;
    int c = fgetc(file);
    while (c != '\0') {
        if (c == EOF) {
            systems::LoggerSystem::getInstance().systemLognError(std::string(utils::RDR_ERROR_READ_STRING));
            return std::nullopt;
        }

        result += static_cast<char>(c);
        c = fgetc(file);
    }

    return result;
}

std::optional<common::bytecodes::ApicaBytecode> utils::readBytecode(FILE *file) {
    uint64_t result;
    if (fread(&result, sizeof(uint64_t), 1, file) != 1) {
        systems::LoggerSystem::getInstance().systemLognError(std::string(utils::RDR_ERROR_READ_BYTECODE));
        return std::nullopt;
    }
    
    if (result > common::bytecodes::ApicaBytecode::BYTECODE_LAST) {
        systems::LoggerSystem::getInstance().systemLognError(std::string(utils::RDR_ERROR_READ_BYTECODE));
        return std::nullopt;
    }
    
    return static_cast<common::bytecodes::ApicaBytecode>(result);
}

std::optional<common::bytecodes::ApicaSpecificationBytecode> utils::readSpecificationBytecode(FILE *file) {
    uint64_t result;
    if (fread(&result, sizeof(uint64_t), 1, file) != 1) {
        systems::LoggerSystem::getInstance().systemLognError(std::string(utils::RDR_ERROR_READ_SPEC_BYTECODE));
        return std::nullopt;
    }
    
    if (result > common::bytecodes::ApicaSpecificationBytecode::SPECIFICATION_LAST) {
        systems::LoggerSystem::getInstance().systemLognError(std::string(utils::RDR_ERROR_READ_SPEC_BYTECODE));
        return std::nullopt;
    }
    
    return static_cast<common::bytecodes::ApicaSpecificationBytecode>(result);
}

std::optional<common::bytecodes::ApicaEntrypointBytecode> utils::readEntryBytecode(FILE *file) {
    uint64_t result;
    if (fread(&result, sizeof(uint64_t), 1, file) != 1) {
        systems::LoggerSystem::getInstance().systemLognError(std::string(utils::RDR_ERROR_READ_ENTRY_BYTECODE));
        return std::nullopt;
    }
    
    if (result > common::bytecodes::ApicaEntrypointBytecode::ENTRYPOINT_LAST) {
        systems::LoggerSystem::getInstance().systemLognError(std::string(utils::RDR_ERROR_READ_ENTRY_BYTECODE));
        return std::nullopt;
    }
    
    return static_cast<common::bytecodes::ApicaEntrypointBytecode>(result);
}

std::optional<common::bytecodes::ApicaBuiltinFunctionBytecode> utils::readBuiltinFuncBytecode(FILE *file) {
    uint64_t result;
    if (fread(&result, sizeof(uint64_t), 1, file) != 1) {
        systems::LoggerSystem::getInstance().systemLognError(std::string(utils::RDR_ERROR_READ_BUILTIN_FUNC_BYTECODE));
        return std::nullopt;
    }

    if (result > common::bytecodes::ApicaBuiltinFunctionBytecode::BUILTIN_FUNC_LAST) {
        systems::LoggerSystem::getInstance().systemLognError(std::string(utils::RDR_ERROR_READ_BUILTIN_FUNC_BYTECODE));
        return std::nullopt;
    }

    return static_cast<common::bytecodes::ApicaBuiltinFunctionBytecode>(result);
}

std::optional<common::bytecodes::ApicaTypeBytecode> utils::readTypeBytecode(FILE *file) {
    uint64_t result;
    if (fread(&result, sizeof(uint64_t), 1, file) != 1) {
        systems::LoggerSystem::getInstance().systemLognError(std::string(utils::RDR_ERROR_READ_TYPE_BYTECODE));
        return std::nullopt;
    }

    if (result > common::bytecodes::ApicaTypeBytecode::TYPE_LAST) {
        systems::LoggerSystem::getInstance().systemLognError(std::string(utils::RDR_ERROR_READ_TYPE_BYTECODE));
        return std::nullopt;
    }

    return static_cast<common::bytecodes::ApicaTypeBytecode>(result);
}